app/src/main/java/com/sdp/movemeet/Backend/Firebase/Firestore/FirestoreActivityManager.java

```diff
@@ -0,0 +1,29 @@
+package com.sdp.movemeet.Backend.Firebase.Firestore;
+
+import com.google.android.gms.tasks.Task;
+import com.google.firebase.firestore.FirebaseFirestore;
+import com.google.firebase.firestore.QuerySnapshot;
+import com.sdp.movemeet.Activity.Activity;
+import com.sdp.movemeet.Backend.Serialization.BackendSerializer;
+
+public class FirestoreActivityManager extends FirestoreManager<Activity> {
+
+    public final static String ACTIVITIES_COLLECTION = "activities";
+
+    private final FirebaseFirestore db;
+    private final String collection;
+
+
+    public FirestoreActivityManager(FirebaseFirestore db, String collection, BackendSerializer<Activity> serializer) {
+        super(db, collection, serializer);
+        this.db = db;
+        this.collection = collection;
+    }
+
+    @Override
+    public Task<QuerySnapshot> search(String field, Object value) {
+        // TODO: return 50 closest activities and execute query on those
```

Comment:
Can the user choose how many closest activities ?

Reply:
Basically, my thinking is that we should only programmatically fetch activities that are close to the user, to avoid having the device fetch all activities in the database and then sorting them locally. For instance, if a users asks for the activities in a 10km radius to their location, the backend service would first fetch 50/100/200 closest activities and check which among those are inside the radius. Of course, we could also fetch all activities inside the radius instead of a fixed number of them, provided we are able to query that to the backend easily. My concern is that all activities are always fetched as of now, which does not seem like a scalable solution.

Reply:
I agree, but I don't understand why this should happen in the search method. Shouldn't there be another method that sorts the db to only retrieve the closest activities, using the distance calculator ? I guess that's how you want to do it. Also, the way the updater updates the list of activites is by sorting them by creation date, and only taking the newly added activities. I'm not sure to see how we can do this "double" sorting (by distance and only new activities) efficiently. But maybe that's a matter for another time ;)

Reply:
Yep, I agree with Kepler as well.

Concerning filtering activities based on both creation date and distance, I guess this should be feasible, we just have to choose what is the most important criterion (between creation date and distance). I would say filtering activities that are in our radius first and then extracting only the activities that are not finished yet (activity date > current date). Finally we filter out those activities such that they don’t exceed the maximum number of closest activities (50/100/200). In summary we would have 3 passes:
1. Radius
2. Activity date
3. Limit number of activities to display